src/app/clusters/electrical-energy-measurement-server/ElectricalEnergyMeasurementCluster.cpp

```diff
@@ -16,237 +16,356 @@
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    }
+    outValue = mMeasurementData.periodicImported;
+    return CHIP_NO_ERROR;
+}
 
-    if (index == kEmberInvalidEndpointIndex)
+CHIP_ERROR ElectricalEnergyMeasurementCluster::GetPeriodicEnergyExported(Optional<EnergyMeasurementStruct> & outValue) const
+{
+    if (!mFeatureFlags.HasAll(ElectricalEnergyMeasurement::Feature::kPeriodicEnergy,
+                              ElectricalEnergyMeasurement::Feature::kExportedEnergy))
     {
-        return nullptr;
+        ChipLogError(Zcl,
+                     "Electrical Energy Measurement: PeriodicEnergyExported requires kPeriodicEnergy and kExportedEnergy features");
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
     }
+    outValue = mMeasurementData.periodicExported;
+    return CHIP_NO_ERROR;
+}
 
-    if (index >= MATTER_ARRAY_SIZE(gMeasurements))
+CHIP_ERROR ElectricalEnergyMeasurementCluster::GetCumulativeEnergyReset(Optional<CumulativeEnergyResetStruct> & outValue) const
+{
+    if (!mEnabledOptionalAttributes.IsSet(CumulativeEnergyReset::Id))
     {
-        ChipLogError(NotSpecified, "Internal error: invalid/unexpected energy measurement index.");
-        return nullptr;
+        ChipLogError(Zcl, "Electrical Energy Measurement: CumulativeEnergyReset requires kCumulativeEnergy feature");
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
     }
-    return &gMeasurements[index];
+    outValue = mMeasurementData.cumulativeReset;
+    return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR SetMeasurementAccuracy(EndpointId endpointId, const MeasurementAccuracyStruct::Type & accuracy)
+CHIP_ERROR ElectricalEnergyMeasurementCluster::SetMeasurementAccuracy(const MeasurementAccuracyStruct & value)
 {
+    if (ValueChanged(mMeasurementData.measurementAccuracy, value))
+    {
+        mMeasurementData.measurementAccuracy = value;
+        NotifyAttributeChanged(Accuracy::Id);
+    }
+    return CHIP_NO_ERROR;
+}
 
-    MeasurementData * data = MeasurementDataForEndpoint(endpointId);
-    VerifyOrReturnError(data != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+CHIP_ERROR ElectricalEnergyMeasurementCluster::SetCumulativeEnergyImported(const Optional<EnergyMeasurementStruct> & value)
+{
+    if (!mFeatureFlags.HasAll(ElectricalEnergyMeasurement::Feature::kCumulativeEnergy,
+                              ElectricalEnergyMeasurement::Feature::kImportedEnergy))
+    {
+        ChipLogError(
+            Zcl, "Electrical Energy Measurement: CumulativeEnergyImported requires kCumulativeEnergy and kImportedEnergy features");
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    }
 
-    data->measurementAccuracy = accuracy;
+    if (ValueChanged(mMeasurementData.cumulativeImported, value))
+    {
+        mMeasurementData.cumulativeImported = value;
+        NotifyAttributeChanged(CumulativeEnergyImported::Id);
+    }
+    return CHIP_NO_ERROR;
```

Comment:
we should handle event generation in the cluster code. Maybe here and maybe sets should be compound (i.e. the entire mMeasurementDaata and not individual bits)